### ISSUE TITLE
Async native dart code?

### ISSUE THREAD

**Issue by duzenko** [dart-lang/sdk#45798]
AFAIK most if not all async functions in Dart/Flutter have to do with I/O and native platform calls.

What if we need to run a very long operation written in native dart? E.g. process image pixels?

I know of isolates, but it has an overhead of serializing inputs and outputs, as well as programmer's man-hours to wrap it up when dealing with existing/external code.

I'm thinking about something like this
```
  void longcode() {
    for(very long loop) {
      // do something
      ...
      // yield CPU to other tasks
      sync_yield();
    }
  }
  ...
  await Future.fromSync(longcode);
```
`sync_yield` could check if the sync code has run over X msec since the last yield and then allow the waiting code to execute before returning.

In short, a native implementation of https://en.wikipedia.org/wiki/Cooperative_multitasking

**Comment by lrhn** [comment]
If you let other code in the same isolate run, then your code *is* asynchronous. It has all the risks and costs of asynchronous code, but none of the affordances, so it's not clear what you want to achieve by not making `longcode` async.

Your `longcode` code could be written as:
```dart
Future<void> longcode() async {
   var sw = Stopwatch()..start();
   for (very long loop) {
     // do something
     ...
     // Yield control 
     if (sw.elapsedMilliseconds > X) {
       await Future(()=>null); // or `Future.microtask(() => null)` if you don't to give timers time to run.
       sw.reset();
     }
  } 
  ...
  await longcode();
```
That's very much the entire *purpose* of asynchronous code, to let other code run while you're not.

**Comment by duzenko** [comment]
Thanks, that's what I was looking for
Here's the test code I'm leaving here in case anyone else will need something like that

longcode1: 10M iter/sec
longcode2: 700K iter/sec
longcode2: 3M iter/sec

```
import 'dart:async';

void longcode1() {
  final d1 = DateTime.now();
  int count = 0;
  while (DateTime.now().difference(d1).inMilliseconds < 1000) {
    count++;
  }
  print(count);
}

Future<void> longcode2() async {
  final d1 = DateTime.now();
  int count = 0;
  while (DateTime.now().difference(d1).inMilliseconds < 1000) {
    count++;
    await Future(() => null);
  }
  print(count);
}

Future<void> longcode3() async {
  final d1 = DateTime.now();
  int count = 0;
  while (DateTime.now().difference(d1).inMilliseconds < 1000) {
    count++;
    await Future.microtask(() => null);
  }
  print(count);
}

Future<Null> main() async {
  print('start');
  final timer = Timer.periodic(Duration(milliseconds: 100), idleTimer);
  switch (3) {
    case 0:
      await Future.delayed(Duration(seconds: 1));
      break;
    case 1:
      longcode1();
      break;
    case 2:
      await longcode2();
      break;
    case 3:
      await longcode3();
      break;
  }
  timer.cancel();
  print('finish');
}

void idleTimer(Timer _) {
  print('timer');
}
```
